rfcs/modal-component.md

```diff
@@ -0,0 +1,169 @@
+# Modal Component
+
+### Exports
+
+* `Modal`
+* `ModalHeader`
+* `ModalBody`
+* `ModalFooter`
+* `StyledRoot`
+* `StyledBackdrop`
+* `StyledDialog`
+* `StyledDialogInner`
+* `StyledHeader`
+* `StyledBody`
+* `StyledFooter`
+* `StyledFooterAction`
+* `StyledClose`
+* `SIZE`
+* `ROLE`
+
+### <Modal/> API
+
+* `children: React.node | ({close: fn}) => React.node` - Required
+  Modal content. The children-as-function API may be preferable for performance reasons (wont render until opened) and also gives users access to a `close()` callback that they can add to buttons
+* `isOpen: boolean` - Required
+  Controls whether the modal is visible or not
+* `closeable: boolean` - Optional, Defaults to `true`
```

Comment:
the modal is always closable. So little confusing. We can combine custom action and default actions (close button, X, backdrop) to allow user to choose. Action can be provided with default close handler and user can just call it, or do some other things in it.

Reply:
There may be use cases where the user *must* take an action before they can close the modal–for example, authenticating before accessing a page. By setting closeable to false, this removes the close button, and prevents backdrop from being clickable.

Reply:
correct. I mentioned my code example, where if we click on X or backdrop nothing happens by default, but only if user does custom action (async or not) and calls default `close` method intentionally whenever user needs. That saves some extra properties and logic

Reply:
ahhh okay I missed your code example. here it is for context:

```
onCloseClick(close, source) => {
   switch (source) {
     case 'ESCAPE': close(result);
     case 'BACKDROP': //do something else or nothing
   }
}
```

I agree, instead of having onBackdropClick, onEscapePress, etc, we can just have keep the single `onClose` and pass a source attribute.

That said, we still need a way to hide the close button, right? No point in showing it if clicking it does nothing. That's what `closeable` is for.

Reply:
ok, let's leave it for this purpose